rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,508 @@
+
+
+#### Examples
+
+The following diagram shows `GET_TELEMETRY (0x00)` request from the server to SDK:
+
+```
+Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
+Transmission Control Protocol, Src Port: 8091, Dst Port: 52370, Seq: 209, Ack: 177535, Len: 3
+WebSocket
+    1... .... = Fin: True
+    .000 .... = Reserved: 0x0
+    .... 0010 = Opcode: Binary (2)
+    0... .... = Mask: False
+    .000 0001 = Payload length: 1
+    Payload
+Data (1 byte)
+    Data: 00
+    [Length: 1]
+
+0000  00 00 00 00 00 00 00 00 00 00 00 00 08 00 45 00   ..............E.
+0010  00 37 20 59 40 00 40 06 1c 66 7f 00 00 01 7f 00   .7 Y@.@..f......
+0020  00 01 1f 9b cc 92 1d 1b b6 e1 52 f6 c7 0d 80 18   ..........R.....
+0030  02 e5 fe 2b 00 00 01 01 08 0a 3d 00 3d 05 3d 00   ...+......=.=.=.
+0040  39 1c 82 01 00                                    9....
+```
+
+And below the response with status `SUCCESS (0x00)` from SDK to the server (the
+payload is truncated, and included as unmasked):
+
+```
+Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
+Transmission Control Protocol, Src Port: 52370, Dst Port: 8091, Seq: 224959, Ack: 212, Len: 41169
+WebSocket
+    1... .... = Fin: True
+    .000 .... = Reserved: 0x0
+    .... 0010 = Opcode: Binary (2)
+    1... .... = Mask: True
+    .111 1111 = Payload length: 127 Extended Payload Length (64 bits)
+    Extended Payload length (64 bits): 88579
+    Masking-Key: d176c493
+    Masked payload
+    Payload
+Data (88579 bytes)
+    [Length: 88579]
+
+Unmasked data (88579 bytes):
+00000  00 73 64 6b 5f 6b 76 5f 72 5f 74 6f 74 61 6c 7b   .sdk_kv_r_total{
+00010  61 67 65 6e 74 3d 22 72 75 62 79 2f 31 2e 32 2e   agent="ruby/1.2.
+00020  33 20 28 52 75 62 79 2f 33 2e 33 2e 35 29 22 2c   3 (Ruby/3.3.5)",
+00030  62 75 63 6b 65 74 3d 22 64 65 66 61 75 6c 74 22   bucket="default"
+00040  2c 68 6f 73 74 3d 22 65 78 61 6d 70 6c 65 2e 63   ,host="example.c
+00050  6f 6d 22 7d 20 31 34 33 39 32 20 31 37 33 31 33   om"} 14392 17313
+00060  31 34 30 35 37 39 35 39 0a 73 64 6b 5f 71 75 65   14057959.sdk_que
+00070  72 79 5f 72 5f 74 6f 74 61 6c 7b 61 67 65 6e 74   ry_r_total{agent
+00080  3d 22 72 75 62 79 2f 31 2e 32 2e 33 20 28 52 75   ="ruby/1.2.3 (Ru
+00090  62 79 2f 33 2e 33 2e 35 29 22 2c 62 75 63 6b 65   by/3.3.5)",bucke
+000a0  74 3d 22 64 65 66 61 75 6c 74 22 2c 68 6f 73 74   t="default",host
+000b0  3d 22 65 78 61 6d 70 6c 65 2e 63 6f 6d 22 7d 20   ="example.com"}
+...
+```
+
+
+## Integration with Custom Meters
+
+When the user might overrides the default meter with their own (e.g. providing
+OpenTelemetry meter), the reporting features, described in this RFC must not be
+disabled and the SDK must still collect and report metrics regarding
+application and library health.
+
+## Configuration
+
+This SDK does not require any additional configuration properties with
+Committed status. Although for the faster adoption and easier testing, it
+should implement the following options and document them as Volatile.
+
```

Comment:
Do we need a `disable_app_telemetry` option?

Reply:
Yep, on second thought, since this will probably be configurable in the server this is not as important as I originally thought.

Reply:
@vishaldhiman22 Do you have any thoughts on this from a PM perspective?

Reply:
It might still be useful to have it be an SDK option -  a user may want to have more control than all or none of their applications to be reporting telemetry, for whatever reason.

Reply:
I agree, it needs to be disable-able.  It can be enabled by default, but I think there needs to be a way to disable at the client.  I guess I differ from @avsej 's comment above.  I consider this a requirement.

When disabled, we should also aim to avoid the overhead associated with this feature.  I know there is some small amount of overhead that will always be there, but the memory and lookup of the counters can be avoided.

Reply:
I will add option to disable the feature, and clarify that app telemetry should not be collected if the endpoint is not specified.